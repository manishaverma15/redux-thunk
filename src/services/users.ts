import React from 'react';

export type User = {
  id: string;
  name: string;
  email: string;
  phoneNumber: string;
}

const BASE_URL = 'api/users';

export async function getUsers(): Promise<User[]> {
  try {
    const response = await fetch(BASE_URL);
    if (!response.ok) {
      throw new Error(`Error: ${response.status} - ${response.statusText}`);
    }

    const data = await response.json();
    return data.map((user: any) => ({
      id: user.id,
      name: user.name,
      email: user.email,
      phoneNumber: user.phoneNumber,
    }));
  } catch (error) {
    console.error('Error fetching users:', error);
    throw error;
  }
}

export async function postUser(postData: any): Promise<User> {
  try {
    const response = await fetch(BASE_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(postData),
    });

    if (!response.ok) {
      throw new Error(`Error: ${response.status} - ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error posting user:', error);
    throw error;
  }
}
