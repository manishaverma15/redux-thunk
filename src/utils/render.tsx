import { ReactElement } from 'react';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { configureStore, PreloadedState, EnhancedStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import userReducer from '../features/Form/userSlice';
import { RootState } from '../store';

interface ExtendedRenderOptions extends RenderOptions {
  preloadedState?: PreloadedState<RootState>;
  store?: EnhancedStore;
}

interface WrapperProps {
  children?: React.ReactNode;
}

function render(
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore({ reducer: { users: userReducer }, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: WrapperProps) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export default render;