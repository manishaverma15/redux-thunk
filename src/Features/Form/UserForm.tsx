import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import './Form.css';
import { AppDispatch } from '../../store';
import { User } from '../../services/users';
import { updateUser, deleteUser, getUsers, postUser } from './userSlice';
import UserTable from './UserTable';

const UserForm: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const users = useSelector((state: any) => state.users.data);
  const [isEditing, setIsEditing] = useState(false);

  const [formData, setFormData] = useState<User>({
    id: uuidv4(),
    name: '',
    email: '',
    phoneNumber: '',
  })

  useEffect(() => {
    dispatch(getUsers());
  }, []);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value,
    })
  }

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    if (isEditing) {
      dispatch(updateUser(formData));
    }
    else {
      dispatch(postUser(formData));
    }

    setIsEditing(false)

    setFormData({
      id: uuidv4(),
      name: '',
      email: '',
      phoneNumber: ''
    })
  };

  const handleDelete = (id: any) => {
    dispatch(deleteUser(id))
  }

  const handleEdit = (id: any) => {
    if (users) {
      const selectedUser = users.find((user: User) => user.id === id)
      console.log('selected-user', selectedUser)
      if (selectedUser) {
        setFormData(selectedUser)
      }
      setIsEditing(true);
    }
  }

  return (
    <>
      <div>
        <h1 style={{ textAlign: 'center' }}>React-Redux Form</h1>
      </div>
      <form onSubmit={handleSubmit}>
        <label>
          Name:
          <input
            type='text'
            name='name'
            value={formData.name}
            onChange={handleChange}
          />
        </label>
        <br />
        <label>
          Email:
          <input
            type='email'
            name='email'
            value={formData.email}
            onChange={handleChange}
          />
        </label>
        <br />
        <label>
          Phone Number:
          <input
            type='number'
            name='phoneNumber'
            value={formData.phoneNumber}
            onChange={handleChange}
          />
        </label>
        <br />
        <button type='submit'>{isEditing ? 'Update User' : 'Add User'}</button>
      </form>

      <UserTable handleEdit={handleEdit} handleDelete={handleDelete} />
    </>
  )
}

export default UserForm;