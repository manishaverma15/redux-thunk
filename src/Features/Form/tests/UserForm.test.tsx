import React from 'react';
import { waitFor, screen, fireEvent } from '@testing-library/react';
import UserForm from '../UserForm';
import render from '../../../utils/render';
import { getUsers, postUser } from '../../../services/users';

jest.mock('../../../services/users');

describe('Form Component', () => {

  beforeAll(() => {
    jest.clearAllMocks();
  })

  const users = [{ id: 'test', name: 'John Roy', email: 'john@gmail.com', phoneNumber: '1234567890' }];

  it('render correctly when there are users', async () => {
    getUsers.mockReturnValue(users);
    const { asFragment } = render(<UserForm />);

    await waitFor(() => screen.getByTestId('user-table'));

    expect(screen.getByText('John Roy')).toBeInTheDocument();
    expect(screen.getByText('john@gmail.com')).toBeInTheDocument();
    expect(screen.getByText('1234567890')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('render correctly when there no users', async () => {
    getUsers.mockReturnValue([]);
    const { asFragment } = render(<UserForm />);

    await waitFor(() => screen.getByTestId('user-table'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders when form is submitted correctly and values are added to the table', async () => {
    const mockUser = { id: 'test1', name: 'user', email: 'newuser@gmail.com', phoneNumber: '9876543210' }
    getUsers.mockReturnValue(users);
    postUser.mockReturnValue(mockUser);

    render(<UserForm />);

    await waitFor(() => screen.getByTestId('user-table'));

    fireEvent.change(screen.getByTestId('name-input'), { target: { value: mockUser.name } });
    fireEvent.change(screen.getByTestId('email-input'), { target: { value: mockUser.email } });
    fireEvent.change(screen.getByTestId('phone-input'), { target: { value: mockUser.phoneNumber } });

    fireEvent.submit(screen.getByTestId('submit-button'));

    await waitFor(() => {
      expect(screen.getByText('user')).toBeInTheDocument();
      expect(screen.getByText('newuser@gmail.com')).toBeInTheDocument();
      expect(screen.getByText('9876543210')).toBeInTheDocument();
    });

    await waitFor(() => screen.getByTestId('user-table'));

    fireEvent.click(screen.getByTestId(`delete-button-${mockUser.id}`));
    await waitFor(() => {
      expect(screen.queryByText('user')).not.toBeInTheDocument();
    });
  });

  it('render when user is edited', async () => {
    const mockUser = { id: 'test', name: 'John Roy', email: 'john@gmail.com', phoneNumber: '1234567890' }
    getUsers.mockReturnValue(users);

    render(<UserForm />);

    await waitFor(() => screen.getByTestId('user-table'));

    fireEvent.click(screen.getByTestId(`edit-button-${mockUser.id}`));

    await waitFor(() => {
      expect(screen.getByTestId('name-input')).toHaveValue(mockUser.name);
      expect(screen.getByTestId('email-input')).toHaveValue(mockUser.email);
      expect(screen.getByTestId('phone-input')).toHaveValue(mockUser.phoneNumber);
    });

    fireEvent.change(screen.getByTestId('name-input'), { target: { value: 'roy' } });
    fireEvent.change(screen.getByTestId('email-input'), { target: { value: 'roy123@gmail.com' } });
    fireEvent.change(screen.getByTestId('phone-input'), { target: { value: '9876543210' } });

    fireEvent.submit(screen.getByTestId('submit-button'));

    fireEvent.change(screen.getByTestId('name-input'), { target: { value: mockUser.name } });
    fireEvent.change(screen.getByTestId('email-input'), { target: { value: mockUser.email } });
    fireEvent.change(screen.getByTestId('phone-input'), { target: { value: mockUser.phoneNumber } });

    await waitFor(() => {
      expect(screen.getByText('roy')).toBeInTheDocument();
      expect(screen.getByText('roy123@gmail.com')).toBeInTheDocument();
      expect(screen.getByText('9876543210')).toBeInTheDocument();
    });

    await waitFor(() => screen.getByTestId('user-table'));
  })

});