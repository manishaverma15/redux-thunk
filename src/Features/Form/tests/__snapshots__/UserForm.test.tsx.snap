// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component render correctly when there are users 1`] = `
<DocumentFragment>
  <div>
    <h1
      style="text-align: center;"
    >
      React-Redux Form
    </h1>
  </div>
  <form>
    <label
      for="name"
    >
      Name:
      <input
        data-testid="name-input"
        name="name"
        placeholder="Name"
        type="text"
        value=""
      />
    </label>
    <br />
    <label
      for="name"
    >
      Email:
      <input
        data-testid="email-input"
        name="email"
        type="email"
        value=""
      />
    </label>
    <br />
    <label
      for="name"
    >
      Phone Number:
      <input
        data-testid="phone-input"
        name="phoneNumber"
        value=""
      />
    </label>
    <br />
    <button
      data-testid="submit-button"
      type="submit"
    >
      Add User
    </button>
  </form>
  <div
    class="user-table"
  >
    <h2>
      Submitted Users
    </h2>
    <table
      data-testid="user-table"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Email
          </th>
          <th>
            Phone Number
          </th>
          <th>
            Edit
          </th>
          <th>
            Delete
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            John Roy
          </td>
          <td>
            john@gmail.com
          </td>
          <td>
            1234567890
          </td>
          <td>
            <button
              data-testid="edit-button-test"
            >
              Edit
            </button>
          </td>
          <td>
            <button
              data-testid="delete-button-test"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Form Component render correctly when there no users 1`] = `
<DocumentFragment>
  <div>
    <h1
      style="text-align: center;"
    >
      React-Redux Form
    </h1>
  </div>
  <form>
    <label
      for="name"
    >
      Name:
      <input
        data-testid="name-input"
        name="name"
        placeholder="Name"
        type="text"
        value=""
      />
    </label>
    <br />
    <label
      for="name"
    >
      Email:
      <input
        data-testid="email-input"
        name="email"
        type="email"
        value=""
      />
    </label>
    <br />
    <label
      for="name"
    >
      Phone Number:
      <input
        data-testid="phone-input"
        name="phoneNumber"
        value=""
      />
    </label>
    <br />
    <button
      data-testid="submit-button"
      type="submit"
    >
      Add User
    </button>
  </form>
  <div
    class="user-table"
  >
    <h2>
      Submitted Users
    </h2>
    <table
      data-testid="user-table"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Email
          </th>
          <th>
            Phone Number
          </th>
          <th>
            Edit
          </th>
          <th>
            Delete
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</DocumentFragment>
`;
